ultravox/data/datasets.py

```diff
@@ -1060,43 +1068,75 @@ def create_dataset(name: str, args: VoiceDatasetArgs) -> data.IterableDataset:
-    """Interleaves multiple IterableDataset objects."""
+    """Interleaves multiple IterableDataset objects based on normalized weights."""
 
     def __init__(
-        self, datasets: Sequence[data.IterableDataset], repeat: bool = False
+        self,
+        datasets: Sequence[data.IterableDataset],
+        stop_strategy: str = dataset_config.InterleaveStopStrategy.last_exhausted,
+        seed: Optional[int] = 42,
+        static: bool = False,
     ) -> None:
         """
         Args:
-            datasets: a list of IterableDataset objects
-            repeat: whether to repeat the datasets indefinitely.
-                This matters most when the datasets have different lengths.
-                Let's say you have two datasets, A and B which have 5 and 3 samples respectively.
-
-                `repeat=False`: [A0, B0, A1, B1, A2, B2, A3, A4]
-                `repeat=True` : [A0, B0, A1, B1, A2, B2, A3, B0, A4, B1, A0, ...]
-
-                NOTE: with `repeat=True`, `__iter__` never stops.
+            datasets: A list of data.IterableDataset objects.
+            stop_strategy: Strategy for stopping iteration. Choose from 'first_exhausted', 'last_exhausted', 'never_stop'.
+            seed: Optional seed for reproducibility.
+            static: If true, the datasets are interleaved in a static order with equal weights.
         """
         super().__init__()
         self._datasets = datasets
-        self._repeat = repeat
+        self._rng = np.random.default_rng(seed)
+        self._static = static
+
+        self._stop_strategy = stop_strategy
+
+        weights = [getattr(ds, "weight", 1) for ds in datasets]
+        total_weight = sum(weights)
+        self._normalized_probs = [w / total_weight for w in weights]
 
     def __iter__(self):
+        # If no datasets are provided, return an empty iterator
+        if not self._datasets:
+            return
+
         iters = [iter(ds) for ds in self._datasets]
-        iter_index = 0
+        exhausted = [False] * len(iters)
+
+        if self._static:
+            static_iter = itertools.cycle(range(len(self._datasets)))
+
+        while True:
+            if self._static:
+                iter_index = next(static_iter)
+            else:
+                iter_index = self._rng.choice(len(iters), p=self._normalized_probs)
 
-        while len(iters):
-            it = iters[iter_index]
             try:
-                val = next(it)
-                iter_index = (iter_index + 1) % len(iters)
-                yield val
+                yield next(iters[iter_index])
             except StopIteration:
-                if not self._repeat:
-                    iters.pop(iter_index)
-                    iter_index %= max(1, len(iters))
-                else:
-                    iters[iter_index] = iter(self._datasets[iter_index])
+                # Reconstruct the iterator
+                iters[iter_index] = iter(self._datasets[iter_index])
```

Comment:
We should only restart the iterator for `last_exhausted` and `never_stop`, I think.

Reply:
Will update to:

```
                exhausted[iter_index] = True

                # Check if stopping condition is met
                if (
                    self._stop_strategy
                    == dataset_config.InterleaveStopStrategy.first_exhausted
                    and any(exhausted)
                ):
                    break
                if (
                    self._stop_strategy
                    == dataset_config.InterleaveStopStrategy.last_exhausted
                    and all(exhausted)
                ):
                    break

                # Recreate the iterator if stopping condition is not met and yield the next sample
                iters[iter_index] = iter(self._datasets[iter_index])
                yield next(iters[iter_index])
  ```

Reply:
perhaps
```
if self._stop_strategy == StopStrategy.FIRST_EXHAUSTED or (
   self._stop_strategy == StopStrategy.LAST_EXHAUSTED and all(exhausted)):
   break
```